src/DynamoCoreWpf/Utilities/ActionDebouncer.cs

```diff
@@ -0,0 +1,28 @@
+using System;
+using System.Threading;
+using System.Windows.Threading;
+
+namespace Dynamo.Wpf.Utilities
+{
+    internal class ActionDebouncer
+    {
+        Dispatcher _dispatcher;
+        CancellationTokenSource _cts;
+        public ActionDebouncer(Dispatcher dispatcher)
+        {
+            _dispatcher = dispatcher;
+        }
+
+        public void Debounce(int timeout, Action action)
+        {
+            if (_cts != null)
+                _cts.Cancel();
+            _cts = new CancellationTokenSource();
+            System.Threading.Tasks.Task.Delay(timeout, _cts.Token).ContinueWith(t =>
```

Comment:
hmm, unfortunately I find this hard to reason about, it seems to me that the continuation will execute on a thread pool thread, which means that the dispatcher invoke call will be made from another thread. I think you should use `BeginInvoke` to avoid any super hard to reason about deadlocks.

Reply:
Changed to use BeginInvoke. Maybe it could benefit from explicitly setting the DispatcherPriority as well?

Reply:
Invoke and BeginInvoke should both execute on the dispatcher's associated thread (in our case the UI thread).
It should not matter if the dispatcher invoke calls are coming from other threads. Performance could be an issue (thread switching)...
We could also use the run the Continuation on the SynchronizationContext.Current  (which in WPF apps, by default, is always associated with the UI thread)